### DISCUSSION TITLE
Possible to use imgui_md without hello_imgui/immapp?

### DISCUSSION THREAD

**Discussion by bgribble** [pthom/imgui_bundle#228] [Q&A] [answered]
The app I'm building doesn't use any of the "frameworks" (hello_imgui, immapp) but is implemented as a normal imgui app, but in Python. 

I am trying to get the Markdown renderer in `imgui_md` working, but trying to run the font loader always fails:
```
from imgui_bundle import imgui_md as markdown

[...]

        markdown.initialize_markdown()
        font_loader = markdown.get_font_loader_function()
        font_loader()

[   2.805    gui]   File "/home/grib/devel/mfp/.virtual/lib/python3.11/site-packages/mfp/gui/imgui/app_window.py", line 89, in _render_task
[   2.805    gui]     font_loader()
[   2.805    gui] RuntimeError: IM_ASSERT( runner != nullptr )   ---   dpi_aware.cpp:49
```

`dpi_aware.cpp`  seems to be part of `hello_imgui` so I am surprised that I'm anywhere in that code. I have been trying to wind my way back through the bindings to figure out what the actual font loader function is, but I am having a hard time tracking it down. The [sample code](https://github.com/pthom/imgui_bundle/blob/main/bindings/imgui_bundle/demos_python/demo_imgui_md.py)  uses `immapp` and I guess it calls the font loading code in C++ somewhere, but I can't figure out where.

Anybody used the `imgui_md` bindings without any of the app frameworks?

**Comment by bgribble** [comment]
I have tracked this down a bit more. The version of `imgui_bundle` I'm using is 1.3.0, and the included version of `imgui_md`  does have a wrapper helper package called `imgui_md_wrapper` which does include `immapp` headers. So this seems to be expected behavior, at least with this version.

**Reply by pthom** [reply, marked as answer]
Yes , unfortunately it is. The original imgui_md requires a lot of set up before you can use it.

The wrapper I have written, removes this burden: it loads all the required fonts, provides a call back for loading and displaying images, a callback for opening links, a custom renderer for code blocks, etc.

However, it does rely on immapp

I Do not see an easy work around for the moment

**Comment by pthom** [comment]
Hello I decided to give it a more complete look 

This commit (https://github.com/pthom/imgui_bundle/commit/4b3f6606f2366879bd635e304a9bb391f92df0d2) makes it possible to use the markdown renderer outside of hello_imgui.

For an example with a pure python back end, look at:
https://github.com/pthom/imgui_bundle/blob/main/bindings/imgui_bundle/python_backends/examples/example_python_backend_glfw3.py

However, we have to rely on some tricks because the pure Python OpenGL renderer normally does not call imgui.backends.opengl3_new_frame() (which is when the font texture is loaded by the native side of hello imgui)

**Reply by bgribble** [reply]
That's great! I can't use the `glfw`  backend (*) but I think I can adapt that for use with the `SDL2` backend that I am using. 

(*) I didn't keep good notes, but I do a lot of custom key and mouse handling and there was something about the GLFW main loop and key handling that was a problem, I don't remember exactly what.
